core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala

```diff
@@ -332,20 +336,26 @@ trait SubmissionMonitor extends FutureSupport with LazyLogging with RawlsInstrum
     dataAccess.entityQuery.getEntities(workflowsWithOutputs.map { case (workflowRec, outputs) => workflowRec.workflowEntityId }).map(_.toMap)
   }
 
-  def saveWorkspace(dataAccess: DataAccess, updatedEntitiesAndWorkspace: Seq[Either[(Option[Entity], Option[Workspace]), (WorkflowRecord, scala.Seq[AttributeString])]]) = {
+  def saveWorkspace(dataAccess: DataAccess, updatedEntitiesAndWorkspace: Seq[Either[(WorkflowEntityUpdate, Option[Workspace]), (WorkflowRecord, scala.Seq[AttributeString])]]) = {
     //note there is only 1 workspace (may be None if it is not updated) even though it may be updated multiple times so reduce it into 1 update
     val workspaces = updatedEntitiesAndWorkspace.collect { case Left((_, Some(workspace))) => workspace }
     if (workspaces.isEmpty) DBIO.successful(0)
     else dataAccess.workspaceQuery.save(workspaces.reduce((a, b) => a.copy(attributes = a.attributes ++ b.attributes)))
   }
 
-  def saveEntities(dataAccess: DataAccess, workspace: Workspace, updatedEntitiesAndWorkspace: Seq[Either[(Option[Entity], Option[Workspace]), (WorkflowRecord, scala.Seq[AttributeString])]]) = {
-    val entities = updatedEntitiesAndWorkspace.collect { case Left((Some(entity), _)) => entity }
-    if (entities.isEmpty) DBIO.successful(0)
-    else dataAccess.entityQuery.save(SlickWorkspaceContext(workspace), entities)
+  def saveEntities(dataAccess: DataAccess, workspace: Workspace, updatedEntitiesAndWorkspace: Seq[Either[(WorkflowEntityUpdate, Option[Workspace]), (WorkflowRecord, scala.Seq[AttributeString])]])(implicit executionContext: ExecutionContext) = {
+    val entityUpdates = updatedEntitiesAndWorkspace.collect { case Left((entityUpdate, _)) if entityUpdate.upserts.nonEmpty => entityUpdate }
+    if(entityUpdates.isEmpty) {
+       DBIO.successful(())
+    }
+    else {
+      DBIO.sequence(entityUpdates map { entityUpd =>
+        dataAccess.entityQuery.saveEntityDeltas(SlickWorkspaceContext(workspace), entityUpd.entityRef, entityUpd.upserts, Seq())
```

Comment:
I don't see `saveEntityDeltas` ever being called with deletes. Do we ever do that?

Reply:
I am not sure the deal with optional outputs, if they even exist in FireCloud. If they do, the lesson we learned the hard way in Firehose is that if
a) an optional output is configured in the method config AND
b) the optional output is not produced in by a workflow AND
c) there is already a value for the attribute indicated by the method config THEN
delete that value. This prevents stale values from hanging around on the entity.

Reply:
Pretty sure we don't have optional outputs.

Reply:
Confirmed with Cromwellians.